File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        vector<int>adj[n+1];
        for(int i=0;i<n;++i){
            if(manager[i]==-1)continue;
            adj[manager[i]].push_back(i);
        }
        int time = informTime[headID];
        queue<pair<int,int>>q;
        q.push({informTime[headID],headID});
        
        while(!q.empty()){
            auto top = q.front();
            q.pop();
            int t = top.first;
            int head = top.second;
            
            for(auto subor:adj[head]){
                time = max(time,t+informTime[subor]);
                q.push({informTime[subor]+t,subor});
            }
        }
        return time;
    }
};

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.java
class Solution {
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        if(n == 1)return 0;
        
        int[] time = new int[n];
        List<List<Integer>> edge = new ArrayList<>();
        
        for(int i = 0;i < n;i++)edge.add(new ArrayList<>());
        
        for(int i = 0;i < n;i++){
            if(manager[i] != -1)
                edge.get(manager[i]).add(i);
        }
        
        int max = 0;
        Queue<Integer> q = new LinkedList<>();
        
        q.add(headID);
        
        while(!q.isEmpty()){
            int curr = q.poll();
            
            for(int sub : edge.get(curr)){
                time[sub] = time[curr] + informTime[curr];
                max = Math.max(max, time[sub]);
                q.add(sub);
            }
        }
        return max;
    }
}